### PR TITLE
[chopsticks] xcm transfer tests

### PR DESCRIPTION
**Pull request by vanderian**

### CODE REVIEW

**Comment by vanderian** [comment]
using latest yarn locally & the modifications might break compilation on some envs

File: test/xcm/statemine.transfer.test.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    let statemine: StatemineNode
+
+    beforeAll(async () => {
+        // devops/chopsticks/statemine.yml port
+        statemine = await StatemineNode.create("ws://127.0.0.1:9948", ChainId.Statemine)
+        await setupApi()
+        setupUsers()
+    })
+
+    // todo repeat for every other asset
```
Comment (goncer) [inline test/xcm/statemine.transfer.test.ts:20]: test.each? 
```
  test.each([address_1, address_2])(
    //, address_1 ])(
    "xyk-pallet: Create new users with bonded amounts.",
    async (address) => {
 ```

File: test/xcm/statemine.transfer.test.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    it("send USDt to mangata and back", async () => {
+        // const op = statemine.xTokenTransfer(ChainId.Mg, AssetId.USDt, AssetId.USDt.unit.mul(BN_TEN), alice)
```
Comment (goncer) [inline test/xcm/statemine.transfer.test.ts:22]: why this commented code?

File: utils/eventListeners.ts [hunk 3]
```diff
@@ -136,7 +136,15 @@ export const waitForEvent = async (
...
-): Promise<void> => {
+): Promise<any[]> => {
+  return (await waitForEvents(api, method, blocks))[0]
+}
+
+export const waitForEvents = async (
+  api: ApiPromise,
+  method: string,
+  blocks: number = 10
+): Promise<any[][]> => {
```
Comment (goncer) [inline utils/eventListeners.ts:147]: Optional: let's try to avoid "any" return type plz!

File: test/xcm/bifrost.transfer.test.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { Balance } from "@polkadot/types/interfaces";
+import { BN, BN_FIVE, BN_TEN } from "@polkadot/util";
+import _ from "lodash";
+import { AssetId, ChainId } from "../../utils/ChainSpecs";
+import { getEvents, waitForEvent, waitForEvents } from "../../utils/eventListeners";
+import { XcmNode } from "../../utils/Framework/Node/XcmNode";
+import { alice, api, setupApi, setupUsers } from "../../utils/setup";
+import { signSendFinalized, signSendSuccess } from "../../utils/sign";
+import { XToken } from "../../utils/xToken";
+
```
Comment (goncer) [inline test/xcm/bifrost.transfer.test.ts:11]: plz add a group to this test! xcm perhaps?

File: test/xcm/statemine.transfer.test.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import { BN, BN_FIVE, BN_TEN } from "@polkadot/util";
+import _ from "lodash";
+import { AssetId, ChainId } from "../../utils/ChainSpecs";
+import { waitForEvent, waitForEvents } from "../../utils/eventListeners";
+import { StatemineNode } from "../../utils/Framework/Node/StatemineNode";
+import { alice, api, setupApi, setupUsers } from "../../utils/setup";
+import { signSendSuccess } from "../../utils/sign";
+import { XToken } from "../../utils/xToken";
+
```
Comment (goncer) [inline test/xcm/statemine.transfer.test.ts:9]: plz add a group to this test! xcm perhaps?

**Review by goncer — APPROVED** [review]

File: test/xcm/statemine.transfer.test.ts (hunk 1 above)
Comment (vanderian) [reply test/xcm/statemine.transfer.test.ts:20]: maybe, also it is needed to modify chostick config yml for alice to have the tokens

File: test/xcm/statemine.transfer.test.ts (hunk 2 above)
Comment (vanderian) [reply test/xcm/statemine.transfer.test.ts:22]: testing xcm for not yet opened channels on kusama

File: utils/eventListeners.ts (hunk 3 above)
Comment (vanderian) [reply utils/eventListeners.ts:147]: object would be better ?

File: utils/Framework/Node/StatemineNode.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  api: ApiPromise;
+  chain: ChainSpec;
+
+  xTokenTransfer(
+    toChain: ChainId,
+    assetId: AssetSpec,
+    amount: BN,
+    toUser: User
+  ): any {
+    expect(ChainSpecs.has(toChain));
```
Comment (goncer) [inline utils/Framework/Node/StatemineNode.ts:22]: expects should be only for test expectations, perhaps we should use hee "asserts" ? 
I can fix them.

Comment (goncer) [reply utils/Framework/Node/StatemineNode.ts:22]: https://blog.logrocket.com/assertion-functions-typescript/

File: utils/Framework/XcmHelper.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  wasmOverride,
+  db,
+  types,
+  localPort,
+  buildBlockMode,
+}: SetupOption): Promise<ApiContext> => {
+  // random port
+  const port = localPort
+    ? localPort
+    : Math.floor(Math.random() * 10000) + 10000;
```
Comment (goncer) [inline utils/Framework/XcmHelper.ts:49]: Lets avoid flakiness by using soem fancy npm package, 
https://www.npmjs.com/package/find-free-port

File: utils/Framework/XcmNetworks.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import dotenv from "dotenv";
+
+import { setupContext, SetupOption } from "./XcmHelper";
+
+dotenv.config();
+
+const endpoints = {
```
Comment (goncer) [inline utils/Framework/XcmNetworks.ts:7]: question: 
why do we need this info here and in the .yml ?

File: utils/Framework/XcmNetworks.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    }),
+  biforst: (options?: Partial<SetupOption>) =>
+    setupContext({
+      wasmOverride: process.env.BIFROST_WASM,
+      blockNumber: toNumber(process.env.BIFROST_BLOCK_NUMBER),
+      endpoint: process.env.BIFROST_ENDPOINT ?? endpoints.bifrost,
+      db: process.env.DB_PATH,
+      ...options,
+    }),
+  mangata: (options?: Partial<SetupOption>) =>
```
Comment (goncer) [inline utils/Framework/XcmNetworks.ts:57]: lets see how long we can suvive without the mangata real types 
:)

File: utils/eventListeners.ts [hunk 10]
```diff
@@ -201,3 +201,130 @@ export const waitForRewards = async (
...
+  return expect(toHuman(codec));
+};
+
+export const expectJson = (codec: CodecOrArray) => {
+  return expect(toJson(codec));
+};
+
+export const expectHex = (codec: CodecOrArray) => {
+  return expect(toHex(codec));
+};
```
Comment (goncer) [inline utils/eventListeners.ts:242]: those expects are test expectations, perhaps we should move them to the utils/setupJest.ts expects

File: package.json [hunk 11]
```diff
@@ -7,12 +7,13 @@
...
   "private": true,
   "dependencies": {
+    "@acala-network/chopsticks": "^0.4.1",
     "@babel/plugin-proposal-decorators": "^7.17.8",
     "@commitlint/config-conventional": "^17.4.2",
     "@mangata-finance/sdk": "1.16.1",
     "@types/bs58": "^4.0.1",
     "@types/jsonpath": "^0.2.0",
-    "@types/node": "^14.14.10",
+    "@types/node": "^18.14.6",
```
Comment (goncer) [inline package.json:16]: does this mean that we need to bump node version everywhere?

**Review by goncer — COMMENTED** [review]
Plz let me know if the suggestions are fine, I can do them :)

File: utils/Framework/XcmHelper.ts (hunk 7 above)
Comment (vanderian) [reply utils/Framework/XcmHelper.ts:49]: ok

File: package.json (hunk 11 above)
Comment (vanderian) [reply package.json:16]: hmm, haven't noticed why it did upgrade, will try to revert

**Review by goncer — APPROVED** [review]

**Pull request merged**